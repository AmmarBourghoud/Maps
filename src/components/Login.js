import React, { useState } from 'react';
import { Avatar, Button, CssBaseline, TextField, FormControlLabel, Checkbox, Link, Box, Typography, Container } from '@material-ui/core';
import {Alert} from '@material-ui/lab';
import { makeStyles } from '@material-ui/core/styles';
import logonamR from '../assets/img/logo-namR-2021.svg';
import { LOGIN_MAIL_ADDRESS, LOGIN_PASSWORD } from '../consts/consts';
//import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

function Copyright() {
  return (
    <Typography variant="body2" color="textSecondary" align="center">
      {'Copyright Â© '}
      <Link color="inherit" href="https://github.com/AmmarBourghoud/namR" target="_blank">
        abourgho
      </Link>{' '}
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: theme.spacing(15),
    height: theme.spacing(15),
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
  alert: {
    root: {
      width: '100%',
      '& > * + *': {
        marginTop: theme.spacing(2),
      }
  },
 },
}));

export default function Login({ setToken }) {
  const classes = useStyles();
  
  const [mail, setMail] = useState("");
  const [password, setPassword] = useState("");
  const [isError, setIsError] = useState(false);
  //const history = useHistory();
  
  const checkCreditentials = (e) => {
    e.preventDefault();
    const token = {
      mail,
      password
    };
    (mail === LOGIN_MAIL_ADDRESS && password === LOGIN_PASSWORD) ? 
        setToken(token) : 
        setIsError(true)
  }

  function showAlert() { 
    return( 
      <div className={classes.alert}>
        <Alert variant="filled" severity="error" onClose={() => setIsError(false)}>
            Mot de passe ou adresse mail invalide !
        </Alert>
      </div>
    );
  }

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
      { isError && showAlert() }
        <Avatar className={classes.avatar} alt="namR logo" src={logonamR} />
        <form className={classes.form} noValidate onSubmit={checkCreditentials}>
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email Address : admin@admin.com"
            name="email"
            autoComplete="email" 
            autoFocus
            value={mail}
            onChange={(e) => setMail(e.target.value)}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password : admin"
            type="password"
            id="password"
            autoComplete="current-password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
          <FormControlLabel
            control={<Checkbox value="remember" color="primary" />}
            label="Remember me"
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            Se connecter
          </Button>
        </form>
      </div>
      <Box mt={8}>
        <Copyright />
      </Box>
    </Container>
  );
}

Login.propTypes = {
  setToken: PropTypes.func.isRequired
}