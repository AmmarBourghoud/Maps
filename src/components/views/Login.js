import React, { useState } from 'react';
import { Avatar, Button, CssBaseline, TextField, FormControlLabel, Checkbox, Link, Box, Typography, Container } from '@material-ui/core';
import {Alert} from '@material-ui/lab';
import { makeStyles } from '@material-ui/core/styles';
import logonamR from '../../assets/img/logo-namR-2021.svg';
import { LOGIN_MAIL_ADDRESS, LOGIN_PASSWORD } from '../../consts/consts';
import PropTypes from 'prop-types';

function Copyright() {
  return (
    <Typography variant="body2" color="textSecondary" align="center">
      {'Copyright Â© '}
      <Link color="inherit" href="https://github.com/AmmarBourghoud/namR" target="_blank">
        abourgho
      </Link>{' '}
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(2),
    backgroundColor: 'white',
    width: theme.spacing(15),
    height: theme.spacing(15),
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
  alert: {
    root: {
      width: '100%',
      '& > * + *': {
        marginTop: theme.spacing(2),
      }
  },
 },
}));

/** 
    * Login component gets the token state's methode from parent component * 
    * Has three local states : *
    * mail : to check the mail through checkCreditentials function *
    * password : to check the password through checkCreditentials function *
    * isError : to alert the user through showAlert function *
**/ 
export default function Login({ setToken }) {
  const classes = useStyles();
  
  const [mail, setMail] = useState("");
  const [password, setPassword] = useState("");
  const [isError, setIsError] = useState(false);
  
  /** 
    * checkCreditentials function * 
    * Check the mail/password entred by user to either updates the token's state or show alert *
  **/   
  const checkCreditentials = (e) => {
    e.preventDefault();
    const token = {
      mail,
      password
    };
    (mail === LOGIN_MAIL_ADDRESS && password === LOGIN_PASSWORD) ? 
        setToken(token) : 
        setIsError(true)
  }

  /** 
    * showAlert function * 
    * Show an alert on wrong user entries *
  **/ 
  function showAlert() { 
    return( 
      <div className={classes.alert}>
        <Alert variant="filled" severity="error" onClose={() => setIsError(false)}>
            Mot de passe ou adresse mail invalide !
        </Alert>
      </div>
    );
  }

  /** 
    * Returns the form and calls the right functions to update states * 
  **/ 
  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
        <div className={classes.paper}>
          { isError && showAlert() }
          <Avatar className={classes.avatar} alt="namR logo" src={logonamR} />
          <form className={classes.form} noValidate onSubmit={checkCreditentials}>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              id="email"
              label="Email Address : admin@admin.com"
              name="email"
              autoComplete="email" 
              autoFocus
              value={mail}
              onChange={(e) => setMail(e.target.value)}
            />
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password : admin"
              type="password"
              id="password"
              autoComplete="current-password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
            <FormControlLabel
              control={<Checkbox value="remember" color="primary" />}
              label="Remember me"
            />
            <Button
              type="submit"
              fullWidth
              variant="contained"
              color="primary"
              className={classes.submit}
            >
            Se connecter
           </Button>
        </form>
      </div>
      <Box mt={8}>
        <Copyright />
      </Box>
    </Container>
  );
}

Login.propTypes = {
  setToken: PropTypes.func.isRequired
}