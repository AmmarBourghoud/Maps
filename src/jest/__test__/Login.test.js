import { render, cleanup, fireEvent } from '@testing-library/react';
import Login from '../../components/views/Login';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

afterEach(cleanup);

it('renders Login component without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Login/>, div); 
  ReactDOM.unmountComponentAtNode(div);
});

it('renders button correctly', () => {
  const {getByTestId} = render(<Login />);
  expect(getByTestId('button')).toHaveTextContent("Se connecter"); 
});

it('renders email form text field', () => {
  const {getByTestId} = render(<Login />);
  expect(getByTestId('mail')).toHaveTextContent("admin@admin.com"); 
});

it('renders password form text field', () => {
  const {getByTestId} = render(<Login />);
  expect(getByTestId('password')).toHaveTextContent("admin"); 
});

it('matches snapshot', () => {
  const tree = renderer.create(<Login />).toJSON(); 
  expect(tree).toMatchSnapshot(); 
}); 

const setup = () => {
  const utils = render(<Login />)
  const inputMail = utils.getByTestId('mail')
  const inputPassword = utils.getByTestId('password')
  return {
    inputMail,
    inputPassword,
    ...utils,
  }
}

it("should update mail on change", () => {
  const {inputMail} = setup();
  fireEvent.change(inputMail, { target: { mail: 'admin@admin' } })
  expect(inputMail.mail).toBe('admin@admin')
});

it("should update password on change", () => {
  const {inputPassword} = setup();
  fireEvent.change(inputPassword, { target: { password: 'admin' } })
  expect(inputPassword.password).toBe('admin')
});

