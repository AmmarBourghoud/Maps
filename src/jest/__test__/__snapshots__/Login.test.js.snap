// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    className="makeStyles-paper-57"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-58"
    >
      <img
        alt="namR logo"
        className="MuiAvatar-img"
        src="logo-namR-2021.svg"
      />
    </div>
    <form
      className="makeStyles-form-59"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        data-testid="mail"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email Address : admin@admin.com
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="email"
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-62 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-64"
            >
              <span>
                Email Address : admin@admin.com
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        data-testid="password"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password : admin
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-62 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-64"
            >
              <span>
                Password : admin
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-66 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              className="PrivateSwitchBase-input-69"
              data-indeterminate={false}
              onChange={[Function]}
              type="checkbox"
              value="remember"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Remember me
        </span>
      </label>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-60 MuiButton-containedPrimary MuiButton-fullWidth"
        data-testid="button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Se connecter
        </span>
      </button>
    </form>
  </div>
  <div
    className="MuiBox-root MuiBox-root-70"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
    >
      Copyright © 
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
        href="https://github.com/AmmarBourghoud/namR"
        onBlur={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        abourgho
      </a>
       
      2021
      .
    </p>
  </div>
</main>
`;
